basic/model.py

```diff
@@ -68,143 +96,180 @@ def __init__(self, config, scope, rep=True):
-        d_cell = SwitchableDropoutWrapper(cell, self.is_train, input_keep_prob=config.input_keep_prob)
-        x_len = tf.reduce_sum(tf.cast(self.x_mask, 'int32'), 2)  # [N, M]
-        q_len = tf.reduce_sum(tf.cast(self.q_mask, 'int32'), 1)  # [N]
+        cell = BasicLSTMCell(hidden_size, state_is_tuple=True)
+        cell_with_dropout = SwitchableDropoutWrapper(cell, self.is_train, input_keep_prob=config.input_keep_prob)
+        # [batch_size, max_num_sentences]
+        passage_len = tf.reduce_sum(tf.cast(self.passage_mask, 'int32'), 2)
+        # [batch_size]
+        question_len = tf.reduce_sum(tf.cast(self.question_mask, 'int32'), 1)
 
         with tf.variable_scope("prepro"):
-            (fw_u, bw_u), ((_, fw_u_f), (_, bw_u_f)) = bidirectional_dynamic_rnn(d_cell, d_cell, qq, q_len, dtype='float', scope='u1')  # [N, J, d], [N, d]
-            u = tf.concat(2, [fw_u, bw_u])
+            # [batch_size, J, hidden_size], [batch_size, hidden_size]
+            (fw_outputs, bw_outputs), ((_, fw_final_state), (_, bw_final_state)) = bidirectional_dynamic_rnn(cell_with_dropout, cell_with_dropout, embedded_question, question_len, dtype='float', scope='encoded_question')
+            encoded_question = tf.concat(2, [fw_outputs, bw_outputs])
             if config.share_lstm_weights:
                 tf.get_variable_scope().reuse_variables()
-                (fw_h, bw_h), _ = bidirectional_dynamic_rnn(cell, cell, xx, x_len, dtype='float', scope='u1')  # [N, M, JX, 2d]
-                h = tf.concat(3, [fw_h, bw_h])  # [N, M, JX, 2d]
+                # [batch_size, max_num_sentences, max_sentence_size, 2*hidden_size]
+                (fw_outputs, bw_outputs), _ = bidirectional_dynamic_rnn(cell, cell, embedded_passage, passage_len, dtype='float', scope='encoded_question')
+                # [batch_size, max_num_sentences, max_sentence_size, 2*hidden_size]
+                encoded_passage = tf.concat(3, [fw_outputs, bw_outputs])
             else:
-                (fw_h, bw_h), _ = bidirectional_dynamic_rnn(cell, cell, xx, x_len, dtype='float', scope='h1')  # [N, M, JX, 2d]
-                h = tf.concat(3, [fw_h, bw_h])  # [N, M, JX, 2d]
-            self.tensor_dict['u'] = u
-            self.tensor_dict['h'] = h
+                # [batch_size, max_num_sentences, max_sentence_size, 2*hidden_size]
+                (fw_outputs, bw_outputs), _ = bidirectional_dynamic_rnn(cell, cell, embedded_passage, passage_len, dtype='float', scope='encoded_passage')
+                # [batch_size, max_num_sentences, max_sentence_size, 2*hidden_size]
+                encoded_passage = tf.concat(3, [fw_outputs, bw_outputs])
+            self.tensor_dict['encoded_question'] = encoded_question
+            self.tensor_dict['encoded_passage'] = encoded_passage
 
         with tf.variable_scope("main"):
+            # The Attention Flow layer
+            # TODO: NOT SURE WHAT TO CALL P0 here / NOT SURE WHAT IT REPRESENTS.
```

Comment:
Min's `p0` is what I call `final_merged_passage` here: https://github.com/allenai/deep_qa/blob/94177eb084fb70fa198e216a05f77e335d388ea0/deep_qa/models/reading_comprehension/bidirectional_attention.py#L152.